Bottom: 6bbc2f526b97dfafc16e7ec34dc8e4e1ce587f40
Top:    84dde5925d2ae4b35a6fc92a878e89811540430a
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-05-20 22:41:38 -0700

librdmacm: Check that send and recv CQs are different before destroying

ucma_destroy_cqs() destroys both the send and recv CQs if they
are non-null.  If the two CQs are actually the same one, this
results in a crash when trying to destroy the second CQ.  Check
that the CQs are different before destroying the second CQ.

This fixes a crash when using rsockets, which sets the send and
recv CQs to the same CQ.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>


---

diff --git a/src/cma.c b/src/cma.c
index c32803d..9cd34cf 100755
--- a/src/cma.c
+++ b/src/cma.c
@@ -1096,10 +1096,10 @@ static void ucma_destroy_cqs(struct rdma_cm_id *id)
 	if (id->recv_cq_channel)
 		ibv_destroy_comp_channel(id->recv_cq_channel);
 
-	if (id->send_cq)
+	if (id->send_cq && (id->send_cq != id->recv_cq))
 		ibv_destroy_cq(id->send_cq);
 
-	if (id->send_cq_channel)
+	if (id->send_cq_channel && (id->send_cq_channel != id->recv_cq_channel))
 		ibv_destroy_comp_channel(id->send_cq_channel);
 }
